src/kubelet/metric/metric.go

```diff
@@ -18,29 +18,29 @@ import (
 const StatsSummaryPath = "/stats/summary"
 
 // GetMetricsData calls kubelet /stats/summary endpoint and returns unmarshalled response
-func GetMetricsData(c client.HTTPClient) (v1.Summary, error) {
+func GetMetricsData(c client.HTTPClient) (*v1.Summary, error) {
 	resp, err := c.Do(http.MethodGet, StatsSummaryPath)
 	if err != nil {
-		return v1.Summary{}, err
+		return nil, fmt.Errorf("performing GET request to kubelet endpoint: %w", err)
 	}
 	defer resp.Body.Close() // nolint: errcheck
-	if resp.StatusCode != http.StatusOK {
-		return v1.Summary{}, fmt.Errorf("error calling kubelet endpoint. Got status code: %d", resp.StatusCode)
-	}
 
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return v1.Summary{}, fmt.Errorf("error reading the response body of kubelet endpoint. Got error: %v", err.Error())
+		return nil, fmt.Errorf("reading the response body of kubelet endpoint (response code %d): %w", resp.StatusCode, err)
 	}
 
-	var summary = new(v1.Summary)
-	err = json.Unmarshal(body, summary)
-	if err != nil {
-		return v1.Summary{}, fmt.Errorf("error unmarshaling the response body. Got error: %v", err.Error())
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response code from kubelet: %d. Response body: %v", resp.StatusCode, body)
```

Comment:
Why do this check after attempting to read the body? It should be fine to call `Close()` on it without consuming it first.

Reply:
It might additionally be easier on memory for large payloads.

Reply:
Body may include the error message, so I guess it's useful to read it it for possible logging, to make debugging easier.